tests/context_storages/test_functions.py

```diff
@@ -0,0 +1,243 @@
+            requests={0: Message(text="useful message"), i: Message(text="some message")},
+        )
+        sleep(0.001)
+
+    # Setup schema so that all requests will be read from database
+    db.context_schema.requests.subscript = ALL_ITEMS
+
+    # Check database length
+    assert len(db) == 100
+
+    # Check that both misc and requests are read as expected
+    for i in range(1, 101):
+        read_ctx = db[f"{context_id}_{i}"]
+        assert read_ctx.misc[f"key_{i}"] == f"ctx misc value {i}"
+        assert read_ctx.requests[0].text == "useful message"
+
+    # Check clear
+    db.clear()
+    assert len(db) == 0
+
+
+def keys_test(db: DBContextStorage, testing_context: Context, context_id: str):
+    # Fill database with contexts
+    for i in range(1, 11):
+        db[f"{context_id}_{i}"] = Context()
+        sleep(0.001)
+
+    # Add and delete a context
+    db[context_id] = testing_context
+    del db[context_id]
+
+    # Check database keys
+    keys = db.keys()
+    assert len(keys) == 10
+    for i in range(1, 11):
+        assert f"{context_id}_{i}" in keys
+
+
+def single_log_test(db: DBContextStorage, testing_context: Context, context_id: str):
+    # Set only one request to be included into CONTEXTS table
+    db.context_schema.requests.subscript = 1
+
+    # Add new requestgs to context
+    for i in range(1, 10):
+        testing_context.add_request(Message(text=f"new message: {i}"))
+    db[context_id] = testing_context
+
+    # Setup schema so that all requests will be read from database
+    db.context_schema.requests.subscript = ALL_ITEMS
+
+    # Read context and check only the two last context was read - one from LOGS, one from CONTEXT
+    read_context = db[context_id]
+    assert len(read_context.requests) == 2
+    assert read_context.requests[8] == testing_context.requests[8]
+    assert read_context.requests[9] == testing_context.requests[9]
+
+
+simple_test.no_dict = False
+basic_test.no_dict = False
+pipeline_test.no_dict = False
+partial_storage_test.no_dict = False
+midair_subscript_change_test.no_dict = True
+large_misc_test.no_dict = False
+many_ctx_test.no_dict = True
+keys_test.no_dict = False
+single_log_test.no_dict = True
+_TEST_FUNCTIONS = [
+    simple_test,
+    basic_test,
+    pipeline_test,
+    partial_storage_test,
+    midair_subscript_change_test,
+    large_misc_test,
+    many_ctx_test,
+    keys_test,
+    single_log_test,
+]
+
+
+def run_all_functions(db: Union[DBContextStorage, Dict], testing_context: Context, context_id: str):
+    frozen_ctx = testing_context.model_dump_json()
+    for test in _TEST_FUNCTIONS:
+        if isinstance(db, DBContextStorage):
+            db.context_schema.append_single_log = True
+            db.context_schema.duplicate_context_in_logs = False
+            for field_props in [value for value in dict(db.context_schema).values() if isinstance(value, SchemaField)]:
+                field_props.subscript = 3
```

Comment:
I like the solution, but in that case we won't be able to test dict the same way anymore (or it will require some additional code, just like the current solution).